### PR TITLE
feat(cli): Support ARGO_HTTP1 for HTTP/1 CLI requests. Fixes #4394

### LINKED ISSUE #4394: Argo CLI does not work with load balancer (e.g. AWS ALB) that do not support HTTP/2
## Summary 

I followed the [docs](https://argoproj.github.io/argo/access-token/) to use `argo` CLI with the access token.

### API Request

It works with API.

```bash
> curl https://$HOST/api/v1/workflow-templates/$NAMESPACE -H "Authorization: $ARGO_TOKEN"
{"metadata":{"se ...}
```

### CLI Request

It doesn't work with CLI.

```bash
> argo template list -n $NAMESPACE -v -e -k
DEBU[0000] CLI version                                   version="{latest+f8e750d.dirty 2020-09-17T22:11:28Z f8e750de5ebab6f3c494c972889b31ef24c73c9b v2.11.0 dirty go1.15.2 gc darwin/amd64}"
DEBU[0000] Client options                                opts="{{$HOST:443 true true}  0x236c500 <nil> 0x236c560}"
FATA[0000] could not find a token

> argo template list -n $NAMESPACE -v --token $ARGO_TOKEN
DEBU[0000] CLI version                                   version="{latest+f8e750d.dirty 2020-09-17T22:11:28Z f8e750de5ebab6f3c494c972889b31ef24c73c9b v2.11.0 dirty go1.15.2 gc darwin/amd64}"
DEBU[0000] Client options                                opts="{{$HOST:443 false false}  0x236c500 <nil> 0x236c560}"
2020/10/28 00:10:32 rpc error: code = Unavailable desc = connection closed

> argo template list -n $NAMESPACE -v -e --token $ARGO_TOKEN
DEBU[0000] CLI version                                   version="{latest+f8e750d.dirty 2020-09-17T22:11:28Z f8e750de5ebab6f3c494c972889b31ef24c73c9b v2.11.0 dirty go1.15.2 gc darwin/amd64}"
DEBU[0000] Client options                                opts="{{$HOST:443 true false}  0x236c500 <nil> 0x236c560}"
2020/10/28 00:10:43 rpc error: code = Unknown desc = : HTTP status code 464; transport: missing content-type field

> argo template list -n $NAMESPACE -v -e -k --token $ARGO_TOKEN
DEBU[0000] CLI version                                   version="{latest+f8e750d.dirty 2020-09-17T22:11:28Z f8e750de5ebab6f3c494c972889b31ef24c73c9b v2.11.0 dirty go1.15.2 gc darwin/amd64}"
DEBU[0000] Client options                                opts="{{$HOST:443 true true}  0x236c500 <nil> 0x236c560}"
2020/10/28 00:10:51 rpc error: code = Unknown desc = : HTTP status code 464; transport: missing content-type field
```

## Diagnostics

What Kubernetes provider are you using? 

- EKS 1.17.9

What version of Argo Workflows are you running? 
```bash
> argo version
argo: latest+5eebce9.dirty
  BuildDate: 2020-10-19T20:59:29Z
  GitCommit: 5eebce9af4409da9de536f189877542dd88692e0
  GitTreeState: dirty
  GitTag: v2.11.6
  GoVersion: go1.15.3
  Compiler: gc
  Platform: darwin/amd64
```

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>
Fixes #4394

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 

# Changes

This PR introduces a new way to configure the CLI: `ARGO_HTTP1`. This sends messages over HTTP using the HTTP API that Argo Server exposes. 

Refactored `CLISuite` into a single suite that can run in different modes (based on `E2E_MODE`).

### CODE REVIEW

File: test/e2e/fixtures/e2e_suite.go [hunk 1]
```diff
@@ -75,36 +76,37 @@ func (s *E2ESuite) BeforeTest(string, string) {
...
 	}
 
 	for _, r := range resources {
-		err := s.dynamicFor(r).DeleteCollection(foregroundDelete, hasTestLabel)
+		err := s.dynamicFor(r).DeleteCollection(&metav1.DeleteOptions{PropagationPolicy: &background}, hasTestLabel)
+		s.CheckError(err)
+	}
+
+	// delete archived workflows from the archive
+	if s.Persistence.IsEnabled() {
```
Comment (alexec) [inline test/e2e/fixtures/e2e_suite.go:100]: moved from above

Comment (alexec) [reply test/e2e/fixtures/e2e_suite.go:100]: make tests about 25% faster too!

**Comment by alexec** [comment]
@jessesuen I've requested your review of the API change.

**Comment by alexec** [comment]
Reverting to draft to fix the `WorkflowLogs` issue first in own PR.

File: cmd/argo/commands/client/conn.go [hunk 2]
```diff
@@ -59,6 +61,9 @@ func NewAPIClient() (context.Context, apiclient.Client) {
 }
 
 func Namespace() string {
+	if overrides.Context.Namespace != "" {
```
Comment (alexec) [inline cmd/argo/commands/client/conn.go:64]: existing bug

File: cmd/argo/commands/root.go [hunk 3]
```diff
@@ -27,9 +23,65 @@ const (
 // NewCommand returns a new instance of an argo command
 func NewCommand() *cobra.Command {
 	var command = &cobra.Command{
-		Use:     CLIName,
-		Short:   "argo is the command line interface to Argo",
-		Example: fmt.Sprintf(`If you're using the Argo Server (e.g. because you need large workflow support or workflow archive), please read %s.`, help.CLI),
+		Use:   CLIName,
+		Short: "argo is the command line interface to Argo",
+		Long: `
+You can use the CLI in the following modes:
```
Comment (alexec) [inline cmd/argo/commands/root.go:29]: I think a lot of value in these docs. Everyone was confused about usage.

**Comment by alexec** [comment]
Stuck for 20m on "Restore go build cache"

**Comment by alexec** [comment]
`TestFunctionalSuite/TestWorkflowTemplateRefWithExitHandler`

File: util/help/topics.go [hunk 4]
```diff
@@ -1,7 +1,7 @@
 package help
 
 const (
-	root      = "https://github.com/argoproj/argo/blob/master/docs"
+	root      = "hhttps://argoproj.github.io/argo"
```
Comment (simster7) [inline util/help/topics.go:4]: `hhttps`?

File: util/flatten/flatten.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+func toMap(in interface{}) map[string]interface{} {
+	data, _ := json.Marshal(in)
+	out := make(map[string]interface{})
+	_ = json.Unmarshal(data, &out)
+	return out
+}
+
+func f(in map[string]interface{}, out map[string]string, prefix string) {
```
Comment (simster7) [inline util/flatten/flatten.go:16]: `f` as a function name is not very descriptive

File: pkg/apiclient/http1/facade.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	u, err := h.url(method, path, in)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest(method, u.String(), bytes.NewReader(data))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", h.authorization)
+	log.Debugf("curl -X %s -H 'Authorization: ******' -d '%s' '%v'", method, string(data), u)
```
Comment (simster7) [inline pkg/apiclient/http1/facade.go:89]: Remove?

File: pkg/apiclient/argo-server-opts.go [hunk 7]
```diff
@@ -1,7 +1,24 @@
...
+import "fmt"
+
 type ArgoServerOpts struct {
-	// argo server URL
-	URL                        string
+	// argo server host port, must be `host:port`, e.g. localhost:2746
+	URL string
+	// any base path needed (e.g. due to being behind an ingress)
+	Path                       string
 	Secure, InsecureSkipVerify bool
```
Comment (simster7) [inline pkg/apiclient/argo-server-opts.go:10]: Minor: one line for each imo

**Review by simster7 — APPROVED** [review]
Mostly a rubber stamp. I don't know this code very well, so I would recommend @jessesuen or @sarabala1979 also taking a look

**Comment by alexec** [comment]
`TestCLIWithServerOverHTTP1Suite/TestCron/Suspend `

**Review by alexec — COMMENTED** [review]
@jessesuen I'm going to take a bit of a liberty with @simster7 approval and merge this PR. The main controversial thing is the names of the new env vars. `ARGO_HTTP1` and `ARGO_BASH_HREF`. We can change these later if you like.

**Pull request merged**